sched/pthread/pthread_join.c

```diff
@@ -152,90 +152,99 @@ int pthread_join(pthread_t thread, FAR pthread_addr_t *pexit_value)
        * stable for we have to do
        */
 
-      sched_lock();
-      pjoin->crefs++;
-
-      /* Check if the thread is still running.  If not, then things are
-       * simpler.  There are still race conditions to be concerned with.
-       * For example, there could be multiple threads executing in the
-       * 'else' block below when we enter!
-       */
-
-      if (pjoin->terminated)
+      if (!pjoin->detached)
```

Comment:
I see this a bit inconsistent. You added code to return `EINVAL` in `pthread_detach` is case if thread is already detached, but here you do not set `ret` to `EINVAL` is case if trying to join a detached thread. But according to spec `The behavior is undefined if the value specified by the thread argument to pthread_join() does not refer to a joinable thread.`, so it is the same `undefined` and for `pthread_detach`.

Reply:
i use `sudo apt-get install  manpages-posix manpages-posix-dev ` and `man 3posix ptread_detach` ，that suggest `If an implementation detects that the value specified by the thread argument to  pthread_detach()  does  not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error`

Reply:
and,i use `phtread_jion` to wait a `detached thread` it also will wait. `phtread_jion` wait a `detached thread` should return now.

Reply:
Return now success and return now an error makes a difference :)

Reply:
> 现在返回成功和现在返回错误会有所不同:)

https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_join.html

Reply:
Seems like you refer to outdated version. Please use this one: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html

<img width="544" alt="Screenshot 2022-04-07 at 10 06 44" src="https://user-images.githubusercontent.com/17704713/162140250-a2480af6-2d69-498a-8678-d959e2c7acd2.png">

Reply:
> Seems like you refer to outdated version. Please use this one: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html
> 
> <img alt="Screenshot 2022-04-07 at 10 06 44" width="544" src="https://user-images.githubusercontent.com/17704713/162140250-a2480af6-2d69-498a-8678-d959e2c7acd2.png">

yes，it also `If an implementation detects that the value specified by the thread argument to pthread_join() does not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error.`

Reply:
So we should add `else { ret = EINVAL }`?